src/charm.py

```diff
@@ -67,17 +70,28 @@ def _on_config_changed(self, event) -> None:
         self.configure_workload_container(event)
 
     def _report_visits_by_ip_action(self, event: ActionEvent) -> None:
-        """Report requests to nginx by IP and report action result."""
+        """Handle the report-visits-by-ip action."""
+        results = self._report_visits_by_ip(event)
+        event.set_results({"ips": str(results)})
+
+    def _report_visits_by_ip(self, _) -> list[(int, str)]:
+        """Report requests to nginx grouped and ordered by IP and report action result."""
         container = self.unit.get_container(CONTAINER_NAME)
-        process = container.exec(
-            [
-                "/bin/sh",
-                "-c",
-                "awk -vDate=`date -d'now-2 hours' +[%d/+%b/%Y:%H:%M:%S` '{ if ($4 > Date) print $1}' /var/log/nginx/access.log | sort  |uniq -c |sort -n | tail",  # noqa: E501
-            ]
+        log_path = "/var/log/nginx/access.log"
+        with container.pull(log_path) as log_file:
+            list_to_search = log_file.readlines()
+        ip_regex = r"[0-9]+(?:\.[0-9]+){3}"
+        ip_list = []
+        current_datetime = datetime.now()
+        start_datetime = current_datetime - timedelta(minutes=20)
+        for line in list_to_search:
```

Comment:
Here is a proposal for breaking this problem down that should make it easy to read and test, let me know what you think:

* Write a function that accepts a line from the log and determine whether it falls within the timedelta we are looking for, if not filter it out (could be called `filter_time`, maybe you can think of a better name?)
* Write a function that extracts the IP address from a line in a log file, this could return `None` or the IP address (call it `get_ip` or maybe you have a better name?)
* Write a function that takes something that is iterable with all the log lines and return a generator with something like this:
  ```
  lines_within_timedelta = (line for line in log_lines if not filter_time(log_line))
  ip_addresses = (get_ip(line) for line in lines_within_timedelta)
  return filter(None, ip_addresses)
  ```

This means you can easily test the functionality for filtering log lines that are not relevant, functionality for extracting IP addresses and a little bit of higher level testing for multiple log lines.

The other feedback here is that it might be worth writing a custom iterator that stops once the timedelta is exceeded. That way, we don't iterator through all lines in the log file, only those that are within the time delta and then we stop. This assumes that the log file is sorted by time which is commonly the case. It also means that we need to iterate over the log file starting with the most recent line and going backwards, which might require starting from the end of the file depending on how it is written to. This action execution time would currently grow in line with the length of the log file due to the for loop. So, instead of the above `filter_time`, maybe write a function that takes a file name and returns an iterable with all the lines that are within the time delta. Let me know if you would like to do some pair programming on this!